plugins.go

```diff
@@ -228,6 +204,44 @@ func RegisterPlugin(name string, plugin Plugin) {
 	plugins[plugin.ServerType][name] = plugin
 }
 
+// EventType represents the event type to be used with event hooks
+type EventType string
+
+const (
+	// StartupEvent represents a startup event
+	StartupEvent = EventType("startup")
+	// ShutdownEvent represents a shutdown event
+	ShutdownEvent = EventType("shutdown")
+)
+
+// EventHook is a type which holds information about a startup hook plugin.
+type EventHook func(eventType EventType) error
+
+// RegisterEventHook plugs in hook. All the hooks should register themselves
+// and they must have a name.
+func RegisterEventHook(name string, hook EventHook) {
+	if name == "" {
+		panic("event hook must have a name")
+	}
+	if _, dup := eventHooks[name]; dup {
+		panic("hook named " + name + " already registered")
```

Comment:
This reminds me, though is not relevant to this PR directly, the names of these plugins should be prefixed with something, like "hook." or "events." so they don't clutter a global namespace as much. What do you think? (Again, I don't think this involves a code change here, but it will be needed in the dev portal I think.)